packages/expo-modules-core/ios/JSI/EXJavaScriptObject.mm

```diff
@@ -34,33 +34,50 @@ - (nonnull instancetype)initWith:(std::shared_ptr<jsi::Object>)jsObjectPtr
   return _jsObjectPtr.get();
 }
 
-#pragma mark - Subscripting
+#pragma mark - Accessing object properties
 
-- (nullable id)objectForKeyedSubscript:(nonnull NSString *)key
+- (BOOL)hasProperty:(nonnull NSString *)name
 {
-  auto runtime = [_runtime get];
-  auto callInvoker = [_runtime callInvoker];
+  return _jsObjectPtr->hasProperty(*[_runtime get], [name UTF8String]);
+}
 
-  if (runtime && callInvoker) {
-    auto value = _jsObjectPtr->getProperty(*runtime, [key UTF8String]);
-    return expo::convertJSIValueToObjCObject(*runtime, value, callInvoker);
-  }
-  return nil;
+- (nonnull EXJavaScriptValue *)getProperty:(nonnull NSString *)name
+{
+  std::shared_ptr<jsi::Value> value = std::make_shared<jsi::Value>(_jsObjectPtr->getProperty(*[_runtime get], [name UTF8String]));
+  return [[EXJavaScriptValue alloc] initWithRuntime:_runtime value:value];
 }
 
-- (void)setObject:(nullable id)obj forKeyedSubscript:(nonnull NSString *)key
+- (nonnull NSArray<NSString *> *)getPropertyNames
 {
-  auto runtime = [_runtime get];
+  jsi::Runtime *runtime = [_runtime get];
+  jsi::Array propertyNamesArray = _jsObjectPtr->getPropertyNames(*[_runtime get]);
+  return expo::convertJSIArrayToNSArray(*runtime, propertyNamesArray, nullptr);
+}
 
-  if (!runtime) {
-    NSLog(@"Cannot set '%@' property when the EXJavaScript runtime is no longer available.", key);
-    return;
-  }
-  if ([obj isKindOfClass:[EXJavaScriptObject class]]) {
-    _jsObjectPtr->setProperty(*runtime, [key UTF8String], *[obj get]);
-  } else {
-    _jsObjectPtr->setProperty(*runtime, [key UTF8String], expo::convertObjCObjectToJSIValue(*runtime, obj));
-  }
+#pragma mark - Modifying object properties
+
+- (void)setProperty:(nonnull NSString *)name value:(nullable id)value
+{
+  jsi::Value jsiValue = expo::convertObjCObjectToJSIValue(*[_runtime get], value);
+  _jsObjectPtr->setProperty(*[_runtime get], [name UTF8String], jsiValue);
+}
+
+- (void)defineProperty:(nonnull NSString *)name value:(nullable id)value options:(EXJavaScriptObjectPropertyDescriptor)options
+{
+  jsi::Runtime *runtime = [_runtime get];
+  jsi::Object global = runtime->global();
+  jsi::Object objectClass = global.getPropertyAsObject(*runtime, "Object");
+  jsi::Function definePropertyFunction = objectClass.getPropertyAsFunction(*runtime, "defineProperty");
+  jsi::Object descriptor = [self preparePropertyDescriptorWithOptions:options];
+
+  descriptor.setProperty(*runtime, "value", expo::convertObjCObjectToJSIValue(*runtime, value));
+
+  // Object.defineProperty(object, name, descriptor)
```

Comment:
```suggestion
```

Reply:
Actually that comment is intentional 😂 This is what happens under the hood but in JS language. I'll make it more clear though.
```suggestion
  // This call is basically the same as `Object.defineProperty(object, name, descriptor)` in JS
```